be/src/storage/lake/meta_file.cpp

```diff
@@ -56,13 +56,34 @@ void MetaFileBuilder::append_delvec(const DelVectorPtr& delvec, uint32_t segment
 
 void MetaFileBuilder::apply_opwrite(const TxnLogPB_OpWrite& op_write,
                                     const std::map<int, std::string>& replace_segments,
-                                    const std::vector<std::string>& orphan_files) {
+                                    const std::vector<std::string>& orphan_files,
+                                    std::map<std::string, uint64_t>& replace_segments_file_size) {
     auto rowset = _tablet_meta->add_rowsets();
     rowset->CopyFrom(op_write.rowset());
+
+    bool upgrage_from_old_version = rowset->segment_size_size() == 0;
+    std::vector<uint64_t> segments_size;
+    segments_size.reserve(replace_segments.size());
     for (const auto& replace_seg : replace_segments) {
         // when handle partial update, replace old segments with new rewrite segments
         rowset->set_segments(replace_seg.first, replace_seg.second);
+
+        // update new rewrite segments size
+        if (LIKELY(!upgrage_from_old_version)) {
+            rowset->set_segment_size(replace_seg.first, replace_segments_file_size[replace_seg.second]);
+        } else {
+            // rowsetMetaPB don't have segment size when upgrade from old_version, so we can not seg_segment_size directly.
+            // just use the vector to keep segment size in order of index, then travel the vector and add_segment finally.
+            segments_size[replace_seg.first] = replace_segments_file_size[replace_seg.second];
+        }
```

Comment:
```suggestion
```

IMO, we can leave these segments size unfilled in this case. The fact that some segments in a rowset have sizes and some don't can complicate the situation.